src/com/google/enterprise/gsafeed/GsafeedHelper.java

```diff
@@ -51,25 +52,28 @@ public static Gsafeed unmarshalWithDtd(URL url) throws JAXBException,
   public static Gsafeed unmarshalWithDtd(InputStream inputStream)
       throws JAXBException, IOException, ParserConfigurationException,
       SAXException {
-    JAXBContext jaxbContext =
-        JAXBContext.newInstance("com.google.enterprise.gsafeed");
-    Unmarshaller unmarshaller = jaxbContext.createUnmarshaller();
-    DocumentBuilderFactory docBuilderFactory =
-        DocumentBuilderFactory.newInstance();
-    docBuilderFactory.setValidating(true);
-    DocumentBuilder docBuilder = docBuilderFactory.newDocumentBuilder();
-    docBuilder.setEntityResolver(new EntityResolver() {
+    SAXParserFactory saxParserFactory = SAXParserFactory.newInstance();
+    saxParserFactory.setValidating(true);
+    saxParserFactory.setXIncludeAware(false);
+    saxParserFactory.setFeature(XMLConstants.FEATURE_SECURE_PROCESSING, true);
+    saxParserFactory.setFeature(
+        "http://xml.org/sax/features/external-parameter-entities", false);
+    saxParserFactory.setFeature(
+        "http://xml.org/sax/features/external-general-entities", false);
+    XMLReader xmlReader = saxParserFactory.newSAXParser().getXMLReader();
+    xmlReader.setEntityResolver(new EntityResolver() {
         @Override
         public InputSource resolveEntity(String publicId,
             String systemId) throws SAXException, IOException {
-          if (publicId.equals("-//Google//DTD GSA Feeds//EN")) {
+          if (publicId != null
+              && publicId.equals("-//Google//DTD GSA Feeds//EN")) {
             return new InputSource(
                 GsafeedHelper.class.getResource("/gsafeed.dtd").openStream());
           }
-          return null; // DTD is the only expected entity in feed files
+          return new InputSource(new StringReader(""));
```

Comment:
what semantic change does this source change reflect?

Reply:
As I was reading about XXE and so on I learned that returning null lets the parser fall back to its default behavior, while returning an empty string means that the empty string is the value for the entity. Since we have the EntityResolver, it seemed reasonable to make use of it here.